Documentation/postmortems/v3.5-data-inconsistency.md

```diff
@@ -0,0 +1,138 @@
+* Etcd has functional tests designed to detect such problems, however they are unmaintained, flaky and are missing crucial scenarios.
+* Etcd v3.5 release was not qualified as well as previous ones. Older maintainers run manual qualification process that is no longer known or executed.
+* Etcd apply code is so complicated that fixing the data inconsistency took almost 2 weeks and multiple tries (). Fix needed to be so complicated that we needed to develop automatic validation for it (https://github.com/etcd-io/etcd/pull/13885).
+* When fixing the main data inconsistency we have found multiple other edge cases that could lead to data corruption (https://github.com/etcd-io/etcd/issues/13514, https://github.com/etcd-io/etcd/issues/13922, https://github.com/etcd-io/etcd/issues/13937).
+
+### Where we got lucky
+
+* We reproduced the issue using etcd functional only because weird partition setup on workstation. Functional tests store etcd data under `/tmp` usually mounted to in memory filesystem. Problem was reproduced only because one of the maintainers has `/tmp` mounted to standard disk.
+
+## Action items
+
+Action items should directly address items listed in lessons learned. 
+We should double down on things that went well, fix things that went wrong, and stop depending on luck.
+
+Action fall under three types, and we should have at least one item per type. Types:
+* Prevent - Prevent similar issues from occurring. In this case, what testing we should introduce to find data inconsistency issues before release, preventing publishing broken release.
+* Detect - Be more effective in detecting when similar issues occur. In this case, improve mechanism to detect data inconsistency issue so users will be automatically informed.
+* Mitigate - Reduce time to recovery for users. In this case, how we ensure that users are able to quickly fix data inconsistency.
+
+Actions should not be restricted to fixing the immediate issues and also propose long term strategic improvements.
+To reflect this action items should have assigned priority: 
+* P0 - Critical for reliability of the v3.5 release. Should be prioritized this over all other work and backported to v3.5.
+* P1 - Important for long term success of the project. Blocks v3.6 release.
+* P2 - Stretch  goals that would be nice to have for v3.6, however should not be blocking.
+
+| Action Item                                                                         | Type     | Priority | Bug                                          |
+|-------------------------------------------------------------------------------------|----------|----------|----------------------------------------------|
+| Etcd testing can reproduce historical data inconsistency issues                     | Prevent  | P0       |                                              |
+| Etcd detects data corruption by default                                             | Detect   | P0       |                                              |
```

Comment:
+ Etcd corruption check is linearizable/negotiates common revision
+ The recovery procedures are documented and tested:
      - bootstaping from other member's snapshot
      - bootstaping from backup
      - bootstraping from backup + WAL log 
+ Snapshots (exchanged between the leader and lagging member) are checked for consistency

Reply:
I tried to generalize `Etcd corruption check is linearizable/negotiates common revision` as `etcd can reliably detect data corruption`. 

I'm not sure how `Snapshots (exchanged between the leader and lagging member) are checked for consistency` would relate to improve data corruption. I haven't seen this discussed, could you expand on it/file an issue?

Reply:
Done: https://github.com/etcd-io/etcd/issues/13973